### ISSUE TITLE
Higher level FFI abstraction

### ISSUE THREAD

**Issue by multimeric** [jorgecarleitao/arrow2#853]
A certain other Arrow crate [has a lovely method on the `Array` that returns two pointers to FFI Array and Schema](https://docs.rs/arrow/latest/arrow/array/trait.Array.html#method.to_raw). Currently arrow2 does not have this, so users have to delve into the ugliness of creating a `Box` and coercing it into a pointer. In practise all of that stuff is just boilerplate, so it would be nicer to be part of `arrow2` to avoid users writing and rewriting it whenever they actually want to do Arrow interop.

**Comment by multimeric** [comment]
If that is the case, could we not provide a method that still does this, but takes two pointers as arguments?

**Comment by jorgecarleitao** [comment]
That is a good idea. In [this example](https://github.com/jorgecarleitao/arrow2/blob/main/examples/ffi.rs#L7) we tried to illustrate this. Maybe that function should be moved to the crate?

Also, note how the field's name and nullability is not really needed; we only need the data_type.

**Comment by multimeric** [comment]
Yes, making them part of the public API would be quite helpful. But also there's still the issue of the array allocation for import into Rust, ie this stuff:
https://github.com/jorgecarleitao/arrow2/blob/c012c9d3095b03877bd5f610d4db746b728e9cba/examples/ffi.rs#L29-L37

I wonder if we could also create a function that just does this, and returns a tuple of `(array_ptr, schema_ptr)` which users can then pass into the function provided by other arrow implementations?

**Comment by multimeric** [comment]
I don't follow your whole argument, but it seems that you are mostly interested in keeping this API so that programmers can convince themselves that the export is safe. I don't particular subscribe to this philosophy, because we are already blindly trusting a lot of things about any given crate. I would rather the unsafe part be largely wrapped in solid library code rather than forcing the intellectual burden and also some safely risk onto the user.

**Comment by zirconium-n** [comment]
`I would rather the unsafe part be largely wrapped in solid library code `

While I agree with this statement, but in this case, you can't warp the whole unsafe part in library code, (the `Box::from_raw` part). Blindly trusting 3rd-party crates is one thing, having to know the internal of a 3rd-party crate is another thing, especially relying an invariant the crates does not guarantee. I would argue warp half unsafe block into library code will _increase_ the intellectual burden.

Besides safety concerns, passing a pointer increases flexibility. This allows the pointer not comes from a `Box<T>` but a `Box<[T]>` or `Arc<T>`, which are totally valid use cases.

**Comment by dbr** [comment]
> I would rather the unsafe part be largely wrapped in solid library code rather than forcing the intellectual burden and also some safely risk onto the user.

I think #629 may help here? It basically describes a high-level, safe wrapper around the FFI between arrow and arrow2
